provider/core/src/resource.rs

```diff
@@ -332,15 +431,22 @@ impl ResourceOptions {
     ///     variant: Some(Cow::Borrowed("GBP")),
     ///     langid: Some(langid!("pt_BR")),
     /// };
-    /// let components = resc_options.get_components();
+    /// let components: Vec<String> = resc_options
+    ///     .iter_components()
+    ///     .map(|s| s.into_owned())
+    ///     .collect();
     ///
     /// assert_eq!(
     ///     ["GBP", "pt-BR"],
-    ///     components.iter().collect::<Vec<&str>>()[..]
+    ///     components[..]
     /// );
     /// ```
-    pub fn get_components(&self) -> ResourceOptionsComponents {
-        self.into()
+    pub fn iter_components(&self) -> impl Iterator<Item = Cow<str>> {
```

Comment:
This is only ever used to extend a `PathBuf`. How about not constructing iterators and cows and instead doing something like:
```rust
pub fn push_to(&self, path_buf: &mut PathBuf) {
  path_buf.push(*self.variant);
  match self.langid {
    Some(langid) => path_buf.push(langid.to_string()),
    _ => {}
  }
}
```

We could do something similar for `ResourceKey` (instead of exposing `get_path` and `iter_components`).

Reply:
Hmm.  Good observation.  I agree that we should clean this up.

One issue is that `std::path::PathBuf` is not `no_std`.  Currently these APIs are `no_std` and are used indirectly by other `no_std` code.

I filed #1516 to follow up.